[infra-proxy-service] iam project organisation mapping & policies

vsingh-msys opened the pull request:
### :nut_and_bolt: Description: What code changed, and why?

<!-- /!\ Please ensure that you are NOT disclosing any customer information without their consent /!\ -->
- Add projects to add organisation
- Manage IAM based on projects and policies.
- ~~Add custom policies to provide access to admins only.~~
- ~~Add migrations script to resolve confilts with ingest policies wildcard with `infra:*`~~

### :chains: Related Resources
https://github.com/chef/automate/issues/1544
[Infra proxy service authz service changes discussion](https://docs.google.com/document/d/1G7Qva6GGLxzdo09oMd7RRB6QEKJAusV-8F1AGz3pSpE/edit?usp=sharing)

### :+1: Definition of Done [Updated on 30 Mar 2020]

 - Creating an organization now allows for including IAM projects.
 - All API calls for organizations (create, update, get, list, delete) now include IAM projects in their output.

### :athletic_shoe: How to Build and Test the Change
- `rebuild components/infra-proxy-service`
- `rebuild components/automate-deployment`
- `rebuld components/automate-gateway`

Steps to test the change: [25-Mar-20]
- Navigate to `cd components/infra-proxy-service` 
- `make setup_docker_pg`
- `make test`

API Update to manage the org with projects.

- `export TOK=$(get_admin_token)`
- Add the server
```
curl -sSkH "api-token: $TOK" -X POST -d '{"name": "SERVER_NAME", "description": "SERVER_DESCRPTION", "fqdn": "example.com", "ip_address": "127.0.0.1"}' https://a2-dev.test/api/v0/infra/servers
```

- Add the project
```
jo id='new_project' name='new project' | \
    curl -sSkH "api-token: $TOK" --data @- \
    https://a2-dev.test/apis/iam/v2/projects?pretty
```

- Create the Org
```
# with projects
curl -sSkH "api-token: $TOK" -X POST -d '{"name": "ORG_NAME1", "admin_user": "ADMIN_USER", "admin_key": "KEY", "projects": ["new_project"]}' https://a2-dev.test/api/v0/infra/servers/REPLACE_WITH_CREATED_SERVER_ID/orgs?pretty

#without projects
curl -sSkH "api-token: $TOK" -X POST -d '{"name": "ORG_NAME2", "admin_user": "ADMIN_USER", "admin_key": "KEY"}' https://a2-dev.test/api/v0/infra/servers/REPLACE_WITH_CREATED_SERVER_ID/orgs?pretty
```
- Update the created org

```
curl -sSkH "api-token: $TOK" -X PUT -d '{"name": "ORG_NAME1", "admin_user": "ADMIN_USER", "admin_key": "KEY", "projects": []}' https://a2-dev.test/api/v0/infra/servers/REPLACE_WITH_SERVER_ID/orgs/REPLACE_WITH_ORG_ID?pretty

```

### :white_check_mark: Checklist

- [x] I have read the [CONTRIBUTING document](https://github.com/chef/automate/blob/master/CONTRIBUTING.md).
- [ ] Tests added/updated?
- [ ] Docs added/updated?
- [x] All commits have been signed-off for the [Developer Certification of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### :camera: Screenshots, if applicable
Signed-off-by: Vivek Singh <vivek.singh@msystechnologies.com>

NOTE: Updated on March 13, 2020

vsingh-msys commented:
This PR is ready for review it would be good if @chef/auth-team can spend some time on it.

Related resource: https://github.com/chef/automate/issues/3019, https://github.com/chef/automate/issues/2710

Here following things covered in this PR:

1. Updated the iam policies resource `infra:chefServers:*` and actions `infra:chefServers:ACTION`, `infra:chefServerOrgs:ACTION`, and `infra:chefServerOrgCookbooks:ACTION`.

2. Added projects field in orgs table and updated the query accordingly.

3. Added projects filter psql function by taking reference of `teams-service`

4. For now added default projects as `(unassigned)` to avoid breaking existing functionality.

Note: it raises an error if policy action has three-level.

````
[policy] couldn't apply v2 template: invalid IsAuthorizedReq.Action: 'infra:chefServer:orgs:delete' does not match regex pattern "^[a-z][a-zA-Z]*(?::[a-z][a-zA-Z]*){2}$"
````
So in order to avoid it keeping policy action as point 1. eg. `infra:chefServerOrgs:delete`

msorens commented on components/infra-proxy-service/storage/postgres/orgs.go:
```diff
@@ -3,27 +3,48 @@ package postgres
 	uuid "github.com/chef/automate/lib/uuid4"
 )
 
 // StoreOrg saves a org to the DB.
```
```suggestion
// StoreOrg saves an org to the DB.
```

bcmdarroch commented on api/external/infra_proxy/infra_proxy.proto:
```diff
@@ -35,150 +35,150 @@ service InfraProxy {
     option (chef.automate.api.policy).action = "read";
-    option (chef.automate.api.iam.policy).resource = "infra:servers:{server_id}:orgs";
-    option (chef.automate.api.iam.policy).action = "infra:serverOrgs:get";
+    option (chef.automate.api.iam.policy).resource = "infra:chefServers:{server_id}:orgs";
```
since there's no other id in this action, it should be safe to use `id` here:
```suggestion
    option (chef.automate.api.iam.policy).resource = "infra:chefServers:{id}:orgs";
```

bcmdarroch commented on api/external/infra_proxy/infra_proxy.proto:
```diff
@@ -35,150 +35,150 @@ service InfraProxy {
     option (google.api.http).post = "/infra/servers/{server_id}/orgs";
     option (google.api.http).body = "*";
-    option (chef.automate.api.policy).resource = "infra:servers:{server_id}:orgs";
+    option (chef.automate.api.policy).resource = "infra:chefServers:{server_id}:orgs";
```
```suggestion
    option (chef.automate.api.policy).resource = "infra:chefServers:{id}:orgs";
```

msorens commented on components/infra-proxy-service/storage/postgres/orgs.go:
```diff
@@ -84,13 +105,18 @@ func (p *postgres) DeleteOrg(ctx context.Context, orgID uuid.UUID) (storage.Org,
 
+	// ensure we do not pass null projects to db
+	if org.Projects == nil {
+		org.Projects = []string{constants.UnassignedProjectID}
```
Is there something unusual in this service that suggests it should start with `UnassignedProjectID` ?
Prior art suggests this should be:
```suggestion
		org.Projects = []string{}
```

vsingh-msys replied:
Yes, it would be an empty assignment. but for now, I intentionally added it, in order to make it optional to bypass `CHECK (projects <> '{}')` so that currently implemented UI won't break.

let me know if it sounds weird?

msorens replied:
Being an "empty assignment" should not be a problem; other services do that. I guess I am still not understanding why this service should be treating it differently.

vsingh-msys replied:
yeah, that's current let's make it similar so that so it won't be confusing. so created issue for updating existing UI https://github.com/chef/automate/issues/3097

msorens commented on components/infra-proxy-service/storage/postgres/servers.go:
```diff
@@ -3,6 +3,7 @@ package postgres
 import (
 	"context"
 
+	// adapter for database/sql
```
Artifact left over from... something; please delete the comment.

vsingh-msys commented on components/infra-proxy-service/service/service.go:
```diff
@@ -17,12 +19,14 @@ type Service struct {
-func Start(l logger.Logger, migrationsConfig migration.Config, connFactory *secureconn.Factory, secretsClient secrets.SecretsServiceClient) (*Service, error) {
-	p, err := postgres.New(l, migrationsConfig)
+func Start(l logger.Logger, migrationsConfig migration.Config, connFactory *secureconn.Factory, secretsClient secrets.SecretsServiceClient,
+	authzClient authz.SubjectPurgeClient, authzV2PoliciesClient authz_v2.PoliciesClient, authzV2AuthorizationClient authz_v2.AuthorizationClient) (*Service, error) {
```
Yup, thanks for pointing it out :+1:

vsingh-msys commented:
@bcmdarroch & @msorens thanks for ur time.

I will definitely add the test cases(which are pending from so long :smile:) and update you accordingly thanks!

msorens commented:
Thanks for adding the `curl` commands in the PR description. But it still does not really tell me what to expect. Those curl commands seem somewhat like setup instructions. Seems like there should be additional commands that fetch data to show projects are working.

In a nutshell, the PR description should answer these questions:
 * How do I know this code is working correctly?
 * What should I expect to see it do?

vsingh-msys commented on components/infra-proxy-service/server/server_test.go:
```diff
@@ -0,0 +1,127 @@
+	"github.com/chef/automate/components/infra-proxy-service/storage/postgres/migration"
+)
+
+func TestHealthGRPC(t *testing.T) {
```
While running this test case it got failed:

```
Running tool: /usr/local/go/bin/go test -timeout 30s github.com/chef/automate/components/infra-proxy-service/server -run ^(TestHealthGRPC)$

--- FAIL: TestHealthGRPC (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0xa091a0]

goroutine 7 [running]:
testing.tRunner.func1(0xc0003d2500)
	/usr/local/go/src/testing/testing.go:874 +0x3a3
panic(0xaca4a0, 0x11a8950)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/chef/automate/components/infra-proxy-service/server.setupInfraProxyService(0xcbd560, 0xc00003c0f0, 0xc0003d2500, 0xcc8be0, 0xc0000108a8, 0x0, 0x2, 0x0, 0x115aa5c, 0x24, ...)
	/chef-automate/dev/automate/components/infra-proxy-service/server/server_test.go:84 +0x4b0
github.com/chef/automate/components/infra-proxy-service/server.TestHealthGRPC(0xc0003d2500)
	/chef-automate/dev/automate/components/infra-proxy-service/server/server_test.go:34 +0x16b
testing.tRunner(0xc0003d2500, 0xbf1ee8)
	/usr/local/go/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:960 +0x350
FAIL	github.com/chef/automate/components/infra-proxy-service/server	0.013s
FAIL
Error: Tests failed.



```

vsingh-msys replied:
here, definitely I am missing something need someone can look and suggest me some direction?

susanev commented:
@vsingh-msys im sooo sorry! but i just got some new information. we need it to be `infra:infraServers` instead of `chefServers`. im hoping this is the last change i ask of you, thank you!!

vsingh-msys commented:
@susanev Okay no issues thanks!

vsingh-msys commented on api/external/infra_proxy/infra_proxy.proto:
```diff
@@ -39,180 +39,180 @@ service InfraProxy {
   rpc GetServers (infra_proxy.request.GetServers) returns (infra_proxy.response.GetServers) {
     option (google.api.http).get = "/infra/servers";
-    option (chef.automate.api.policy).resource = "infra:servers";
+    option (chef.automate.api.policy).resource = "infra:infraServers";
```
Updated the policy as per the latest update from @susanev

vsingh-msys commented on components/infra-proxy-service/Makefile:
```diff
@@ -23,16 +25,22 @@ build:
 kill_docker_pg:
 	docker rm -f infra-proxy-postgres || true
 
+truncate:
```
Added to truncate the servers table data for testing purposes only.

vsingh-msys commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+
+		t.Run("when a valid org is submitted", func(t *testing.T) {
+			ctx := context.Background()
+			secretsMock.EXPECT().Create(gomock.Any(), &newSecret, gomock.Any()).Return(secretID, nil)
```
At present, we have decided to mock the secrets client. Added the required mock which will be accessed by org operations.

vsingh-msys commented:
@chef/auth-team I have incorporated the review comments and this PR is ready for review.

Most of the test cases I have covered for org's operations in [orgs_test.go](https://github.com/chef/automate/blob/VSingh/infra-proxy-service-iam-projects-mapping/components/infra-proxy-service/server/v1/orgs_test.go)

Additionally, we need to cover `servers` related test cases as well which I planing to cover in further PRs as this PR getting too bulky. 

In order to run the cases:
 - Naviate to `cd components/infra-proxy-service`
 - run `make setup_docker_pg`
 - run `make test`

Let me know further if anything required from my side. Thanks

bcmdarroch commented on .studio/infra-proxy-service:
```diff
@@ -73,17 +73,17 @@ function infra_service_load_sample_data() {
+  chef-automate dev grpcurl automate-gateway -- -d \
   "$(cat << EOF
-    {"name": "$5", "admin_user": "$6", "admin_key": "$7", "server_id": $server_id}
+    {"name": "$5", "admin_user": "$6", "admin_key": "$7", "server_id": $server_id, "projects": ["(unassigned)"]}
```
are you trying to create these so that they're `unassigned` ? In that case you would want to pass an empty projects array. No object in Automate should ever be assigned to `unassigned` - we infer it from the fact that their projects are an empty array.

vsingh-msys replied:
Yes, I was under the impression that we need to assign projects to `unassigned` because I have seen check constraints in `teams-service` unfortunately that was old migration.

Surely will remove this. Thanks

bcmdarroch commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,221 @@
+			cleanupServer(ctx, t, cl, serverRes.Server.Id)
+		})
+
+		t.Run("when the  exists", func(t *testing.T) {
```
missing what `the` is?

bcmdarroch commented on components/infra-proxy-service/storage/postgres/orgs.go:
```diff
@@ -3,53 +3,73 @@ package postgres
+		projects = []string{}
+	}
+
+	// will only return an error if authz is in v2.1 mode
```
this comment doesn't apply anymore since we merged the force-upgrade to v2.1

bcmdarroch commented on components/infra-proxy-service/storage/postgres/orgs.go:
```diff
@@ -3,53 +3,73 @@ package postgres
+	"github.com/lib/pq"
 	"github.com/pkg/errors"
 
+	authz_v2 "github.com/chef/automate/api/interservice/authz/v2"
```
we're about to drop all the `v2` names in authz so you could name this `authz`

bcmdarroch commented on components/infra-proxy-service/storage/postgres/postgres.go:
```diff
@@ -7,23 +7,25 @@ import (
 	"github.com/lib/pq" // adapter for database/sql
 	"github.com/pkg/errors"
 
+	authz_v2 "github.com/chef/automate/api/interservice/authz/v2"
```
```suggestion
	authz "github.com/chef/automate/api/interservice/authz/v2"
```

bcmdarroch commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+	t.Run("CreateOrg", func(t *testing.T) {
+		test.ResetState(ctx, t, serviceRef)
+
+		t.Run("when a valid org is submitted", func(t *testing.T) {
```
i find it easier to read these tests when they include the expected behavior.
```suggestion
		t.Run("when a valid org is submitted, creates the new org successfully", func(t *testing.T) {
```

bcmdarroch commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+	t.Run("GetOrgs", func(t *testing.T) {
+		test.ResetState(context.Background(), t, serviceRef)
+
+		t.Run("when there is no orgs exists", func(t *testing.T) {
```
```suggestion
		t.Run("when no orgs exist", func(t *testing.T) {
```

bcmdarroch commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+			assert.Equal(t, 0, len(resp.GetOrgs()))
+		})
+
+		t.Run("when the project filter is empty returns all orgs", func(t *testing.T) {
```
this test name is great!

msorens commented:
Just an FYI: I updated your steps for "How to Build and Test the Change" because they were not working. (Needed to create a project first, needed lowercase for the project name, made org names distinct so you could run both curls with and without projects, and a couple other items.)

msorens commented:
As I have stated, I am not really sure what the scope of work is for this PR. Please let me know if this accurately characterizes it:

Definition of done:
1. Creating an organization now allows for including IAM projects.
2. All API calls for organizations (create, update, get?, list?) now include IAM projects in their output.

I put question marks on `get` and `list` because you did not mention them but seems like they should be included...?

msorens commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+	secretWithID := newSecret
+	secretWithID.Id = "fake id"
+
+	t.Run("CreateOrg", func(t *testing.T) {
```
Please add some tests where CreateOrg fails. This would include:
(a) for any required field have a test where that field is omitted and check that you get an error; and
(b) if you provide a server id that does not exist or is invalid.

vsingh-msys replied:
Yes, sure will add, I left it behind the bar due to PR is getting bulky.

msorens commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+		})
+	})
+
+	t.Run("GetOrgs", func(t *testing.T) {
```
Excellent set of tests for various project combinations! 👍

msorens commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+		})
+	})
+
+	t.Run("UpdateOrg", func(t *testing.T) {
```
Similar comment to CreateOrg: consider adding tests where an error occurs if required fields are missing or server id is invalid.

vsingh-msys commented on components/infra-proxy-service/server/v1/orgs.go:
```diff
@@ -54,7 +54,7 @@ func (s *Server) CreateOrg(ctx context.Context, req *request.CreateOrg) (*respon
 		return nil, err
```
Sure @msorens this sounds feasible to me. Earlier I was under the impression to have version kind of stuff. Thanks

vsingh-msys commented on components/automate-gateway/handler/infra_proxy/orgs.go:
```diff
@@ -53,13 +53,14 @@ func (a *InfraProxyServer) GetOrgByName(ctx context.Context, r *gwreq.GetOrgByNa
-// CreateOrg posts a org upstream
+// CreateOrg posts an org upstream
 func (a *InfraProxyServer) CreateOrg(ctx context.Context, r *gwreq.CreateOrg) (*gwres.CreateOrg, error) {
 	req := &infra_req.CreateOrg{
```
Yes, At present for both servers and orgs, it is using system-generated GUIDs but while designing the schema I have already taken this into consideration so planing to use human-readable ids as:

 - For servers `name` will be used as ID and `description` would be `name` similar to teams services.

 - For orgs `name` will be used as ID. because it is unique and qualify for human-readable ID.

Will tackle this into further PR.

vsingh-msys replied:
Earlier, we had some discussion for the same and have provided the review comments https://github.com/chef/automate/pull/2209#discussion_r355164907

vsingh-msys commented on components/infra-proxy-service/constants/constants.go:
```diff
@@ -0,0 +1,6 @@
+package constants
+
+// IAM v2 default project ID
```
Yup, my bad will remove v1 & v2 if anywhere in the code.

vsingh-msys commented on components/infra-proxy-service/server/v1/orgs_test.go:
```diff
@@ -0,0 +1,908 @@
+			secretsMock.EXPECT().Delete(gomock.Any(), secretID, gomock.Any())
+
+			req := &request.CreateOrg{
+				Name:      "4thcoffee",
```
Sure this seems to be helpful!

vsingh-msys replied:
Re: Org name `4thcoffee` I am using most of the time, as this is inherited from chef-server docs https://docs.chef.io/install_server/ :smiley: 

Do I need to replace this as well?

vsingh-msys replied:
Renaming for now to clear the offenses `make spell`.

vsingh-msys commented:
> I put question marks on get, list, and delete because you did not mention them but seems like they should be included...?

Thanks @msorens for providing a clean and clear message, just pointed here earlier I have started with create and update projects but later I have covered all operations of orgs with projects. definitely it should include all.

vsingh-msys commented:
As per the suggestions provided by @msorens have covered:

1) Test cases for CreateOrg:

- when the org required field name is missing or empty, raise an invalid argument error
- when the org required field server id is missing or empty, raise an invalid argument error
- when the server does not exist, raise server not found error

2) Test cases for UpdateOrg: 
- when the org Id is missing or empty, raise an invalid argument error
- when the submitted ID is not a UUIDv4, raise an invalid argument error
- when the org required field name is missing or empty, raise an invalid argument error
- when the org required field server id is missing or empty, raise an invalid argument error
- when the server does not exist, raise server not found error

3) Renamed some of the earlier test cases names as per "when condition, then expectation"

4) Added other minor changes, suggestions.

Could you please review the changes again?

Thanks for detailed the review!!!
